WalletWasabi/Blockchain/TransactionBroadcasting/TransactionBroadcaster.cs

```diff
@@ -128,93 +143,71 @@ public async Task<BroadcastingResult> BroadcastAsync(SmartTransaction tx)
-		if (!mempoolService.TryAddToBroadcastStore(tx)) // So we'll reply to INV with this transaction.
-		{
-			Logger.LogDebug($"Transaction {txId} was already present in the broadcast store.");
-		}
+		var entry = mempoolService.GetOrAdd(tx);
 		var invPayload = new InvPayload(tx.Transaction);
 
-		await node.SendMessageAsync(invPayload).WaitAsync(TimeSpan.FromSeconds(3)).ConfigureAwait(false); // ToDo: It's dangerous way to cancel. Implement proper cancellation to NBitcoin!
+		await node.SendMessageAsync(invPayload).WaitAsync(TimeSpan.FromSeconds(3), cancellationToken).ConfigureAwait(false);
 
-		if (mempoolService.TryGetFromBroadcastStore(txId, out TransactionBroadcastEntry? entry))
-		{
-			var broadcastTimeoutTask = Task.Delay(7000);
-			var broadcastFinishedTask = await Task.WhenAny([broadcastTimeoutTask, entry.BroadcastCompleted.Task]).ConfigureAwait(false);
+		var broadcastTimeoutTask = Task.Delay(TimeSpan.FromSeconds(10), cancellationToken);
+		var broadcastTask = entry.BroadcastCompleted.Task;
+		var broadcastFinishedTask = await Task.WhenAny([broadcastTimeoutTask, broadcastTask]).ConfigureAwait(false);
 
-			if (broadcastFinishedTask == broadcastTimeoutTask)
-			{
-				return BroadcastingResult.Fail(new BroadcastError.Unknown($"Timed out to broadcast to {node.RemoteSocketEndpoint} node"));
-			}
-			node.DisconnectAsync("Thank you!");
-			Logger.LogInfo($"Disconnected node: {node.RemoteSocketAddress}. Successfully broadcast transaction: {txId}.");
+		if (broadcastFinishedTask == broadcastTimeoutTask)
+		{
+			return BroadcastingResult.Fail(new BroadcastError.Unknown($"Timed out to broadcast to {node.RemoteSocketEndpoint} node"));
+		}
 
-			var propagationTimeoutTask = Task.Delay(7000);
-			var propagationTask = entry.PropagationConfirmed.Task;
-			var propagationFinishedTask = await Task.WhenAny([ propagationTimeoutTask, propagationTask]).ConfigureAwait(false);
+		var broadcasters = await broadcastTask.ConfigureAwait(false);
+		return BroadcastingResult.Ok(new BroadcastOk.BroadcastByNetwork(broadcasters));
+	}
 
-			if (propagationFinishedTask == propagationTimeoutTask)
-			{
-				return BroadcastingResult.Fail(new BroadcastError.Unknown("Timed out to verify propagation."));
-			}
+	private	async Task<BroadcastingResult> ConfirmPropagationAsync(SmartTransaction tx, CancellationToken cancellationToken)
+	{
+		var entry = mempoolService.GetOrAdd(tx);
+		var propagationTimeoutTask = Task.Delay(TimeSpan.FromSeconds(12), cancellationToken);
+		var propagationTask = entry.PropagationConfirmed.Task;
+		var propagationFinishedTask = await Task.WhenAny([propagationTimeoutTask, propagationTask]).ConfigureAwait(false);
 
-			var propagators = await propagationTask.ConfigureAwait(false);
-			return BroadcastingResult.Ok(new BroadcastOk.BroadcastedByNetwork(propagators));
+		if (propagationFinishedTask == propagationTimeoutTask)
+		{
+			BroadcastingResult.Fail(new BroadcastError.Timeout("Timed out to verify propagation"));
```

Comment:
```suggestion
			return BroadcastingResult.Fail(new BroadcastError.Timeout("Timed out to verify propagation"));
```

Also please relax a bit to 20s.
In many of my tests timeout was happening too fast